### ISSUE TITLE
IndexError: list index out of range

### ISSUE THREAD

**Issue by akshitac8** [ShuLiu1993/PANet#25]
After a few Iteration, I get list index out of range error on
 ```  
mini_kwargs = dict([(k, v[i]) for k, v in kwargs.items()])
```

**Comment by Iamal1** [comment]
Me too when I try to use 4 GPU

**Comment by akshitac8** [comment]
set ``` drop_last = True ```

**Comment by Iamal1** [comment]
But it's set by default.

**Comment by akshitac8** [comment]
Can you post the full error?

**Comment by Iamal1** [comment]
I just checked the logs and I guess it's same as yours.
`mini_kwargs = dict([(k, v[i]) for k, v in kwargs.items()])`
![image](https://user-images.githubusercontent.com/26734206/54256961-afe6e100-4598-11e9-8eb9-bcb27b956746.png)

**Comment by akshitac8** [comment]
it is the same as mine and its because of inconsistent data input to the GPU. check your train_net_step.py file for minibatch and drop_last variables.

**Comment by Iamal1** [comment]
`   maskRCNN = mynn.DataParallel(maskRCNN, cpu_keywords=['im_info', 'roidb'],
                                 minibatch=True)`
```
    batchSampler = BatchSampler(
        sampler=MinibatchSampler(ratio_list, ratio_index),
        batch_size=args.batch_size,
        drop_last=True
    )
```
Actually there are only these two snippets for the two variables, I never modify them, so they are true by default,  So I wonder what's your default setting.

**Comment by Iamal1** [comment]
I figure out what's wrong with my situation. In roi_data.loader, it says this:
```
def collate_minibatch(list_of_blobs):
    """Stack samples seperately and return a list of minibatches
    A batch contains NUM_GPUS minibatches and image size in different minibatch may be different.
    Hence, we need to stack smaples from each minibatch seperately.
    """
    Batch = {key: [] for key in list_of_blobs[0]}
```
I have 8 GPUs but set NUM_GPUS=4 in config.yaml and comment `    cfg.NUM_GPUS = torch.cuda.device_count()` in train_net_step.py to avoid adaptive_setting, so in that line 82 this error occurs, the device_id list has length 8 by default in data_parallel.py, but v only has 4 elements. 
I want to try this because I meet out of memory problem when I use 8 GPUS with batch_size 8 and backbone ResNeXt101. Not familiar with parallel module and this code, I simply hope that the memory from free GPU could be used. But it doesn't work for this code, which by default takes use of all GPUS and separate data on them.
Maybe setting minibatch to False will skip this problem, but I've been training on new machines and never have a try. Anyway, thanks for your help and suggestions.

**Comment by akshitac8** [comment]
Did you use CUDA_VISIBLE_DEVICES? You should not have met with out of memory error bcoz I also run on 8 GPUs with 16 batch size.

**Comment by Iamal1** [comment]
No, if I use CUDA_VISIBLE_DEVICES, it's the same for 4 card and 8 card, both meet OOM. My GPU each only has 12G memory, it could run for a few steps, but at some step I will meet OOM, so I have to decrease my batch size.

**Comment by akshitac8** [comment]
Don't set NUM_GPUS in the yaml file. and reduce the batch size even for all the 8 gpus

**Comment by Iamal1** [comment]
Yes, that's what I do, so I can train the model with R50 and batch size 8, but not batch size 16.

**Comment by jhtao1860** [comment]
@akshitac8  @Iamal1  if you change the NUM_GPUS in .yaml from 4 to 2, you should  modify the code of "the train_net_step.py "  like this：
 maskRCNN = mynn.DataParallel(maskRCNN, device_ids=[0, 1], cpu_keywords=['im_info', 'roidb'],
                                 minibatch=True)
